system/threads.cpp

```diff
@@ -115,44 +166,155 @@ long __syscall_futex(uint32_t* uaddr, int futex_op, ...)
 				if (*uaddr != val)
 				{
 					futexSpinLock.unlock();
-					return EAGAIN;
+					return -EAGAIN;
 				}
 				mainThreadWaitAddress.store(uaddr);
 				if (*uaddr != val)
 				{
 					mainThreadWaitAddress.store(0);
 					futexSpinLock.unlock();
-					return EAGAIN;
+					return -EAGAIN;
 				}
 				futexSpinLock.unlock();
-				int64_t startTime = 0;
-				if (timeout != -1)
+				addNodeToFutexList(&futexNode);
+				while (futexNode.sleepAddress != nullptr)
+				{
+					while (mainThreadWaitAddress.load() != 0)
+					{
+						if (timeout != -1)
+						{
+							struct timespec timeNowStruct;
+							clock_gettime(CLOCK_MONOTONIC, &timeNowStruct);
+							int64_t timeNow = timeNowStruct.tv_sec * 1000000000 + timeNowStruct.tv_nsec;
+							if (timeNow - startTime >= timeout)
+							{
+								removeNodeFromFutexList(&futexNode);
+								return -ETIMEDOUT;
+							}
+						}
+					}
+					futexNode.sleepAddress = futexNode.nextSleepAddress;
+					futexNode.nextSleepAddress = nullptr;
+					futexSpinLock.lock();
+					mainThreadWaitAddress.store(futexNode.sleepAddress);
```

Comment:
why the spinlock here?

Reply:
This was left over from debugging, no need to spinlock here.